kvserver: convert ingestion to batch early

pav-kv opened the pull request:
This refactors `applySnapshotRaftMuLocked` to convert a snapshot ingestion into a write batch early, before generating any additional writes. This serves a few purposes:

1. In the "small snapshot" case, avoids extra work under `raftMu`. Previously, we would first write/sync a bunch of trivial SSTables to disk, and then read them back and convert to a batch. Now, all these extra writes go directly to the batch.
2. The snapshot preparation code (`prepareSnapApply`) becomes simpler, in that it doesn't need to track the redundant "cleared spans" anymore, only to pass them later to the `ConvertFilesToBatchAndCommit` method.
3. When we implement WAG replay, we will no longer need to integrate with `ConvertFilesToBatchAndCommit`. Instead, we will use the standard "batch commit" API.

Related to #155377

pav-kv committed:
kvserver: convert ingestion to batch early

This refactors applySnapshotRaftMuLocked to convert a snapshot ingestion
into a write batch early, before generating any additional writes. This
serves a few purposes:

- In the "small snapshot" case, avoids extra work under raftMu.
  Previously, we would first write/sync a bunch of trivial SSTables to
  disk, and then read them back and convert to a batch. Now, all these
  extra writes go directly to the batch.

- The snapshot preparation code (prepareSnapApply) becomes simpler, in
  that it doesn't need to track the redundant "cleared spans" anymore,
  only to pass them later to the ConvertFilesToBatchAndCommit method.

- When we implement WAG replay, we will no longer need to integrate with
  ConvertFilesToBatchAndCommit. Instead, we will use the standard "batch
  commit" API.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/155532)
<!-- Reviewable:end -->

pav-kv commented:
Need to fix `TestConvertFilesToBatchAndCommit`, but the PR is reviewable.

pav-kv commented on pkg/kv/kvserver/replica_raftstorage.go:
Correction: actually `inSnap.SSTSize` does not change and always reflects only the size of the snapshot. So no change in this behaviour.

pav-kv commented on pkg/storage/pebble.go:
```diff
@@ -2584,11 +2579,7 @@ func (p *Pebble) ConvertFilesToBatchAndCommit(
-		return err
-	}
-	return batch.Commit(true)
+	return nil
```
`return err`

sumeerbhola replied:
reminder
<!-- Reviewable comment -Oc2CdM02Pz-hQBNwNcK:brm6u7r -->
<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> @sumeerbhola reviewed 4 of 6 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @pav-kv)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 613 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2GwbSBIHFYFQUy9kb:-Oc2GwbSBIHFYFQUy9kc:b-ydc4dh) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/kv/kvserver/replica_raftstorage.go#L613)):*
> ```Go
> 		defer batch.Close()
> 		if err := r.store.TODOEngine().IngestAndExciseFilesToWriter(
> 			ctx, inSnap.SSTStorageScratch.SSTs(), exciseSpan, batch,
> ```

I don't think passing `exciseSpan` is sufficient. We have created sstables for the MVCC key space (which will be excised too). All of those contain RANGEDELs for the span they are writing over. That RANGEDEL is not read by `ConvertFilesToBatchAndCommit` which uses an `EngineIterator`. So we need to write those other RANGEDELs to the batch too.
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 619 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2G7Ww4lCHcsxHTlla:-Oc2G7Ww4lCHcsxHTllb:b-x9kkga) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/kv/kvserver/replica_raftstorage.go#L619)):*
> ```Go
> 		// Redirect the additional writes directly into the same batch.
> 		writeSST = func(ctx context.Context, write func(context.Context, storage.Writer) error) error {
> 			return write(ctx, batch)
> ```

I assume we don't have to pay attention to the ordering wrt whether these happen before or after the call to `IngestAndExciseFilesToWriter` since these must not be overlapping with those files as the normal `IngestAndExise` path requires these to be non-overlapping.
___
*[`pkg/kv/kvserver/snapshot_apply_prepare.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2DUQK2GEqwv5YxlC6:-Oc2DUQK2GEqwv5YxlC7:b-cqanc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/kv/kvserver/snapshot_apply_prepare.go#L161)):*
> ```Go
> 					ClearReplicatedByRangeID:   true,
> 					ClearUnreplicatedByRangeID: true,
> 					MustUseClearRange:          s.ingesting,
> ```

I get nervous with changes like this (allowing `MustUseClearRange` to be false) because I don't know how thorough the existing unit testing of these cases is in kvserver land.
___
*[`pkg/storage/engine.go` line 1076 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2BOh_8yXRLYIiEFtd:-Oc2BOha1_f1kmkmsl_c:b-xxr1fw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/storage/engine.go#L1076)):*
> ```Go
> 	// This should be easy since we use NewSSTEngineIterator to read the ssts,
> 	// which supports multiple levels.
> 	IngestAndExciseFilesToWriter(
> ```

let's call this `IngestAndExciseLocalFilesToWriter`, since it makes clear why we require the same contract as `IngestLocalFiles*`. And it keeps the TODO more obviously meaningful since we do not have yet have the capability to support the more general parameters of `IngestAndExciseFiles`.
___
*[`pkg/storage/pebble.go` line 2546 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2CKUa00sqEuRp-vzF:-Oc2CKUa00sqEuRp-vzG:b60h3k8) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba2b8d9ced29cba91a241ffbf23ce8c44f93b80/pkg/storage/pebble.go#L2546)):*
> ```Go
> 
> 	batch := p.NewWriteBatch()
> 	for i := range clearedSpans {
> ```

Presumably the other `clearedSpans` have already been written to `writer`. I don't remember the details of all the things that can be getting cleared (subsumed ranges etc.).


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 613 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2GwbSBIHFYFQUy9kb:-Oc4v4GCDi7QPyiDQ6Uw:bz6k2ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/kv/kvserver/replica_raftstorage.go#L613)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't think passing `exciseSpan` is sufficient. We have created sstables for the MVCC key space (which will be excised too). All of those contain RANGEDELs for the span they are writing over. That RANGEDEL is not read by `ConvertFilesToBatchAndCommit` which uses an `EngineIterator`. So we need to write those other RANGEDELs to the batch too.
</blockquote></details>

Good point. I think we need rangedels for all replicated keyspans [produced](https://github.com/cockroachdb/cockroach/blob/d1cef1a1ff36af682125983fcdbc0d15060cb3f1/pkg/kv/kvserver/replica_raftstorage.go#L314-L320) by the `MultiSSTWriter`.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/kv/kvserver/snapshot_apply_prepare.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2DUQK2GEqwv5YxlC6:-Oc4xV1TBwBcwZ6lsw-w:b4uxpcl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/kv/kvserver/snapshot_apply_prepare.go#L161)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I get nervous with changes like this (allowing `MustUseClearRange` to be false) because I don't know how thorough the existing unit testing of these cases is in kvserver land.
</blockquote></details>

`MustUseClearRange` is being removed in https://github.com/cockroachdb/cockroach/pull/155549, I'll update this PR when done.


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/kv/kvserver/snapshot_apply_prepare.go` line 161 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2DUQK2GEqwv5YxlC6:-OfdSj_C36oRB8OhLmgc:buw48xd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/kv/kvserver/snapshot_apply_prepare.go#L161)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

`MustUseClearRange` is being removed in https://github.com/cockroachdb/cockroach/pull/155549, I'll update this PR when done.
</blockquote></details>

This is removed now.


<!-- Sent from Reviewable.io -->

pav-kv committed:
kvstorage: rm unused cleared spans return value

Epic: none
Release note: none

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155532)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 613 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2GwbSBIHFYFQUy9kb:-Ofe1-y71dlMJbX_j5AO:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/kv/kvserver/replica_raftstorage.go#L613)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Good point. I think we need rangedels for all replicated keyspans [produced](https://github.com/cockroachdb/cockroach/blob/d1cef1a1ff36af682125983fcdbc0d15060cb3f1/pkg/kv/kvserver/replica_raftstorage.go#L314-L320) by the `MultiSSTWriter`.
</blockquote></details>

Fixed
___
*[`pkg/kv/kvserver/replica_raftstorage.go` line 619 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2G7Ww4lCHcsxHTlla:-Ofe5YWT1N8kOD6LwH5J:bnqy1o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/kv/kvserver/replica_raftstorage.go#L619)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I assume we don't have to pay attention to the ordering wrt whether these happen before or after the call to `IngestAndExciseFilesToWriter` since these must not be overlapping with those files as the normal `IngestAndExise` path requires these to be non-overlapping.
</blockquote></details>

Sounds right, we know there is no overlap. In a way, this ordering here is the natural one: first write everything we have so far (the SSTs), and then generate the other bits on top.
___
*[`pkg/storage/engine.go` line 1076 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2BOh_8yXRLYIiEFtd:-Ofe4DqXAXav7KlVvpL6:by0f1zt) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b3675214a3703b1c3850b2ae6189ab1c4324a7a/pkg/storage/engine.go#L1076)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

let's call this `IngestAndExciseLocalFilesToWriter`, since it makes clear why we require the same contract as `IngestLocalFiles*`. And it keeps the TODO more obviously meaningful since we do not have yet have the capability to support the more general parameters of `IngestAndExciseFiles`.
</blockquote></details>

Renamed to `IngestLocalFilesToWriter`, since it looks more like `IngestLocalFiles` to me.

Does the TODO still make sense, or needs tweaking?
___
*[`pkg/storage/pebble.go` line 2546 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Oc2CKUa00sqEuRp-vzF:-Ofe4jHmAVv2mNKTGZRD:bhxahty) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ba2b8d9ced29cba91a241ffbf23ce8c44f93b80/pkg/storage/pebble.go#L2546)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Presumably the other `clearedSpans` have already been written to `writer`. I don't remember the details of all the things that can be getting cleared (subsumed ranges etc.).
</blockquote></details>

Yes, except the "already" bit. Other `clearedSpans` will be written to the `writer` as part of preparing the snapshot (yes, subsumed ranges etc). So we do the "convert to batch" first, and then write other things to this batch (incl. other clears).


<!-- Sent from Reviewable.io -->

pav-kv committed:
kvserver: force ingests in TestStoreRangeMergeRaftSnapshot

This test verifies the content of the SST files produced during the
snapshot application. In the follow-up commit, we stop generating SST
files early, when the snapshot is being applied as a write batch. In
this case, the list of SST files is not complete, which breaks the test.
Fix it by proactively forcing "apply as ingest" path.

Epic: none
Release note: none

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 10 of 10 files at r6, 7 of 7 files at r7, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-:-Og3S23D8_kXcz48meQ4:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @pav-kv)

___
*[`pkg/kv/kvserver/snapshot_apply_prepare_test.go` line 105 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Og3QguV5a2X_2c88kKc:-Og3QguWEv7MmfxK6ovH:b-w3vf3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/92b06b4c0b5964d49f233ff0474c8d04f60d0eb1/pkg/kv/kvserver/snapshot_apply_prepare_test.go#L105)):*
> ```Go
> 	require.NoError(t, err)
> 
> 	for _, span := range rditer.MakeReplicatedKeySpans(swb.desc) {
> ```

I don't understand this part of the test. It presumably is testing what prepareSnapApply does above. But then it generates a bunch of its own output that is based on the `swb.desc` that it provided to `swb`. Is that just to make the output understandable to the reader, and not to test anything?


<!-- Sent from Reviewable.io -->

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155532)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/kv/kvserver/snapshot_apply_prepare_test.go` line 105 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/155532#-Og3QguV5a2X_2c88kKc:-Og3WV03-pvLVb4V_UH_:b-t344sk) ([raw file](https://github.com/cockroachdb/cockroach/blob/92b06b4c0b5964d49f233ff0474c8d04f60d0eb1/pkg/kv/kvserver/snapshot_apply_prepare_test.go#L105)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't understand this part of the test. It presumably is testing what prepareSnapApply does above. But then it generates a bunch of its own output that is based on the `swb.desc` that it provided to `swb`. Is that just to make the output understandable to the reader, and not to test anything?
</blockquote></details>

Yeah, it's not testing any functionality, just informational. It mimics what the snapshot construction code does overall, which is split across `MultiSSTWriter` and `snapWriteBuilder`. The former clears and overrides the replicated spans of the range, while the latter writes additional "fixups" (destroys subsumed replicas if any, and a few more things for our range).

I'll add a comment here before merging.


<!-- Sent from Reviewable.io -->

pav-kv committed:
kvserver: add comment explaining the test

Epic: none
Release note: none
